src/gui/mainwindow.cpp

```diff
@@ -611,34 +611,72 @@ void MainWindow::readSettings()
 
 void MainWindow::balloonClicked()
 {
-    if (isHidden()) {
+    const Preferences* const pref = Preferences::instance();
+    if (baloonShowReason == BALLOON_REASON_TORRENT_FINISHED && pref->openDestinationFolderOnBalloonClick()) {
+        finishedTorrentBalloonClicked();
+    }
+    else {
+        if (isHidden()) {
+            if (ui_locked) {
+                // Ask for UI lock password
+                if (!unlockUI())
+                    return;
+            }
+            show();
+            if (isMinimized())
+                showNormal();
+        }
+
+        raise();
+        activateWindow();
+    }
+}
+
+void MainWindow::finishedTorrentBalloonClicked()
+{
+    const Preferences* const pref = Preferences::instance();
+    if (pref->openDestinationFolderOnBalloonClick() && !balloonClickDestinationPath.isEmpty()) {
         if (ui_locked) {
             // Ask for UI lock password
             if (!unlockUI())
                 return;
         }
-        show();
-        if (isMinimized())
-            showNormal();
+        if (balloonClickDestinationSingleFile) {
+            Utils::Misc::openFolderSelect(balloonClickDestinationPath);
+        }
+        else {
+            Utils::Misc::openPath(balloonClickDestinationPath);
+        }
     }
-
-    raise();
-    activateWindow();
 }
 
-void MainWindow::addTorrentFailed(const QString &error) const
+void MainWindow::addTorrentFailed(const QString &error)
 {
     showNotificationBaloon(tr("Error"), tr("Failed to add torrent: %1").arg(error));
 }
 
 // called when a torrent has finished
-void MainWindow::finishedTorrent(BitTorrent::TorrentHandle *const torrent) const
+void MainWindow::finishedTorrent(BitTorrent::TorrentHandle *const torrent)
```

Comment:
If you removed const-ness only to cache the torrent properties, please consider making those variables mutable instead.